### PR TITLE
Odinbook

### PR DESCRIPTION
**Pull request by aniketrathi**
### Endpoints

- #### Auth
```
User signup
POST /auth
Status:200
Request: {
                firstName,
                lastName,
                email,
                password,
                confirmPassword,
                photo,
            } 
```
```
User Log In -
 POST /auth/login
  Status: 200
  Request: {
                email,
                password
            }
  Response Header: {
                      user-auth-token: <jwt token>
                    }
```

```
Facebook Login - 
POST /facebook
Status:200
Request: {
                access_token,
            }
  Response Header: {
                      facebook-auth-token: <jwt token>
                    }
```

- #### User

```
GET /users
To get All users
Status:200
```

```
GET /users/:userid
To get user of id: userid
Status:200
API called in Profile.js
```
```
PUT /users/:userid
To update user of id: userid
Status:200
Request : {
                firstName,
                lastName,
                email,
                password,
                confirmPassword,
                photo,
}
```
```
DELETE /users/:userid
To delete user of id: userid
Status:200
```
```
GET /users/:userid/posts
To show all the posts of the user
Status:200
API called in Profile.js 
```
```
GET /users/:userid/friendrequests
To show all friend requests of the user
Status:200
API called in friend-request.js
```
```
PUT /users/:userid/friend
To accept the request
Status:200
Request : {
                 user : {
                           friends:[
                                         _id,
                            ]
                   }
}
```
```
PUT /users/:userid/unfriend
To unfriend the user
Status:200
Request : {
                 user : {
                           friends:[
                                         _id,
                            ]
                   }
}
```
```
GET /users/search/:pattern
To search users.
Status: 200
API called in search.js in frontend
```
- #### POST
```
GET /posts
To get all posts
Status: 200
API called in post.js in frontend
```
```
POST /posts
To add the post
Status: 200
Request: {
                author,
                content,
                likes: [],
}
```
```
PUT /posts/:postid
To update the post 
Status: 200
Request : {
                content,
}
```
```
DELETE /post/:postid
To delete the post
Status: 200
```
```
PUT /post/:postid/like
To like the post
Status: 200
Request : {
                 post : {
                           likes:[
                                         _id,
                            ]
                   }
}
```
```
PUT /post/:postid/dislike
To dislike the post if it is already liked by that user.
Status: 200
Request : {
                 post : {
                           likes:[
                                         _id,
                            ]
                   }
}
```
- #### Friend Request
```
POST /friendrequests
To send a friend request
Status: 200
Request : {
                sender,
                receiver,
}
```
```
DELETE /friendrequests/:friendrequestid
To delete the friend request.
Status: 200
```
- #### Comments
```
GET /posts/:postid/comments
To show comments on the particular post
Status: 200
API called in post-detail.js in frontend
```
```
POST /posts/:postid/comments
To add a comment on the particular post
Status: 200
Request : {
                content,
                author, 
                post,
}
```
### Automated test cases added

- #### None

### Manual test cases run

- Checks all the endpoints
- Creates few test data

### CODE REVIEW

File: client/src/components/auth/facebook-login.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+function FacebookLogin() {
+  const { getLoggedIn } = useContext(AuthContext);
+
+  const history = useHistory();
+
+  async function fbLogin(data) {
+    await axios.post(`${process.env.REACT_APP_API_URL}/auth/facebook`, {
+      access_token: data,
+    });
+    console.log(data);
```
Comment (balwant-jtc) [inline client/src/components/auth/facebook-login.js]: remove console logs from every file.

File: client/src/components/posts/Post.js [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      .put(`http://localhost:5000/posts/${post._id}/like`, { _id: user })
+      .then((res) => console.log(res))
+      .catch((err) => console.log(err));
+    window.location.href = "/posts";
+  }
+
+  function dislikeHandler(e) {
+    e.preventDefault();
+    axios
+      .put(`http://localhost:5000/posts/${post._id}/dislike`, { _id: user })
```
Comment (balwant-jtc) [inline client/src/components/posts/Post.js]: HOST can be stored as constant and use that everywhere.

File: server/routes/auth-route.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        .json({ errorMessage: "Wrong username or password!" });
+
+    /// sign the token ///
+    const token = jwt.sign(
+      {
+        user: existingUser._id,
+      },
+      process.env.JWT_SECRET,
+      {
+        expiresIn: 86400,
```
Comment (balwant-jtc) [inline server/routes/auth-route.js]: Ideally configurable things should be in separate constant files

Comment (balwant-jtc) [reply server/routes/auth-route.js]: expiresIn is configurable thing here

File: server/app.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  cors({
+    origin: ["http://localhost:3000"],
+    credentials: true,
+  })
+);
+app.use(passport.initialize());
+app.use(fileUpload());
+
+app.use("/auth", authRouter);
+app.use("/", userRouter);
```
Comment (balwant-jtc) [inline server/app.js]: this is not a way to declare routes, it should be like :
app.use("/users", userRouter);
app.use("/friends", friendRequestRouter);
app.use("/posts", postRouter);
app.use("/comments", commentRouter);

File: server/routes/auth-route.js [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+env.config();
+
+const router = express.Router();
+
+/// SIGNUP ROUTE ///
+router.post("/", async (req, res) => {
+  try {
+    const { email, password, confirmPassword, firstName, lastName } = req.body;
+    var { photo } = req.body;
```
Comment (balwant-jtc) [inline server/routes/auth-route.js]: avoid var, use let or const instead

File: server/routes/user-route.js [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+const check = require("../middlewares/auth-middleware").auth;
+const FriendRequest = require("../models/friend-request");
+const Post = require("../models/post");
+const User = require("../models/user");
+
+const router = express.Router();
+
+router.get("/users", check, (req, res) => {
+  try {
+    User.find()
```
Comment (balwant-jtc) [inline server/routes/user-route.js]: Use async await for better readability in every mongoose query.

File: server/routes/user-route.js [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+        details: ["Update result did not return 1."],
+      });
+    }
+  } catch (err) {
+    console.error(err);
+    res.status(500).send();
+  }
+});
+
+router.get("/users/:userid/posts", check, async (req, res) => {
```
Comment (balwant-jtc) [inline server/routes/user-route.js]: no need to use async here.

File: server/routes/auth-route.js (hunk 3 above)
Comment (aniketrathi) [reply server/routes/auth-route.js]: I stored in env file

File: server/app.js (hunk 4 above)
Comment (aniketrathi) [reply server/app.js]: ok Cool

File: server/routes/auth-route.js (hunk 5 above)
Comment (aniketrathi) [reply server/routes/auth-route.js]: done

File: server/routes/user-route.js (hunk 6 above)
Comment (aniketrathi) [reply server/routes/user-route.js]: ok done

File: server/routes/user-route.js (hunk 7 above)
Comment (aniketrathi) [reply server/routes/user-route.js]: ya removed

File: client/src/components/auth/facebook-login.js (hunk 1 above)
Comment (aniketrathi) [reply client/src/components/auth/facebook-login.js]: done

File: client/src/components/posts/Post.js (hunk 2 above)
Comment (aniketrathi) [reply client/src/components/posts/Post.js]: done

**Pull request merged**